### PR TITLE
feat: dependency cache, configurable timeouts, correct harvest scope

### LINKED ISSUE #23896: [dash-licenses] Cache ClearlyDefined license data to skip already-resolved dependencies
## Is your enhancement related to a problem? Please describe

On every run the tool queries ClearlyDefined for **all** dependencies, regardless of whether they
were already resolved in a previous run. On a project like `eclipse-che/che-dashboard` (~1 421
total dependencies) this means ~1 421 API calls on every invocation — even when `yarn.lock` has
not changed and all licenses are already recorded in `.deps/prod.md` / `.deps/dev.md`.

At the same time, three root causes prevent previously queried results from being reused even
when they exist in the output files:

**Root cause A — uncategorized yarn3 deps never written to output files.**
The yarn3 lockfile parser produces `prod`, `dev`, and `all` arrays. Deps that appear in `all`
but in neither `prod` nor `dev` (graph-traversal gaps for optional/transitive packages) are
approved by the API but never written to `prod.md`/`dev.md`, so they are never cached and are
re-queried on every run.

**Root cause B — transitive deps only written to EXCLUDED with `--harvest`.**
Without `--harvest`, transitive unresolved deps are printed to the console but not persisted to
`EXCLUDED/`. On the next run they are re-queried from scratch.

**Root cause C — `loadResolvedCache` does not read EXCLUDED files.**
Even after Root cause B is fixed, transitive deps stored in `EXCLUDED/prod.md` /
`EXCLUDED/dev.md` are still re-queried because the cache loader only reads `prod.md` and
`dev.md`.

## Describe the solution you'd like

Introduce a resolved-dependency cache that is read at the start of each run. Before querying
ClearlyDefined, load `.deps/prod.md`, `.deps/dev.md`, `EXCLUDED/prod.md`, and `EXCLUDED/dev.md`
into a `Map<identifier, {license, cq}>`. Any dependency whose **Resolved CQs** column is
non-empty is skipped — only truly new or previously unresolved deps are sent to the API.

Result on `eclipse-che/che-dashboard`:

| | API calls |
|---|---|
| Before | ~1 421 (all deps on every run) |
| After | ~93 (1 328 served from cache) |

`--recheck` bypasses the cache and re-queries everything.

Additional improvements included in the same change:

1. **Enriched cache-miss log** — breaks down misses into `prod`, `dev`, `prod+dev` buckets
   and distinguishes version-bumped packages from genuinely new ones.
2. **Configurable POST/GET timeouts** — `--post-timeout` (default 30 000 ms) and
   `--get-timeout` (default 5 000 ms) with one automatic POST retry before falling back to
   individual GETs.
3. **Harvest scoped to direct-unresolved deps only** — previously harvest was triggered for
   every `notfound` result, including transitive deps that belong to `EXCLUDED/`. Harvest now
   runs inside `processAndGenerateDocuments()` and is called only with the identifiers that end
   up in `problems.md`.
4. **EXCLUDED auto-cleanup** — on every run, EXCLUDED entries whose CQ is a `[clearlydefined]`
   link are verified against the API; approved entries are removed from the EXCLUDED file in the
   same run. With `--harvest`, all entries including orphans and `transitive dependency` rows are
   re-checked via the new `recheckAndCleanExcluded()` function.

### Bug 1 — Generated `.deps` files contain hyperlinks from npm registry metadata

`extractLicenseInfo` in `npm/bump-deps.ts` and `yarn3/bump-deps.ts` read `homepage` and
`repository.url` from `node_modules/<pkg>/package.json` and stored the result in `LicenseInfo.URL`.
`arrayToDocument` then wrapped every package name that had a URL in a markdown hyperlink:

```
| [`qs@6.15.2`](https://github.com/ljharb/qs) | BSD-3-Clause | transitive dependency |
| [`axios@1.16.1`](https://axios-http.com)     | MIT           | #28196               |
```

These npm registry URLs are not documentation links and should not appear in the output.

**Fix:** removed `URL` from `LicenseInfo`, removed `homepage`/`repository.url` extraction from
all three package managers (npm, yarn, yarn3), and removed the hyperlink-rendering branch from
`arrayToDocument`. Package names are now always plain backtick code.

### Bug 2 — EXCLUDED packages get a license sourced from `node_modules`

Same `extractLicenseInfo` path read `license` from `node_modules/<pkg>/package.json` even for
packages that are in `EXCLUDED/`. The value was correct by coincidence (e.g. `MIT` for axios)
but originated from the wrong source and would silently diverge if the `node_modules` field
differed from the ClearlyDefined record.

**Fix:** same as Bug 1 — `extractLicenseInfo` no longer reads any fields beyond what is needed
for the license file.

### Bug 3 — `removeUnusedExcludes` skipped linked-name EXCLUDED rows

`PackageManagerUtils.removeUnusedExcludes()` matched only plain `` `pkg@version` `` rows. Linked-name rows
`` [`pkg@version`](url) `` were skipped, so entries logged as "removed" after approval stayed in
the EXCLUDED file.

**Fix:** updated `removeUnusedExcludes` regex to match both formats, and normalized both the row
identifier and the `identifiersToRemove` set through `coordinateToIdentifier()` so Yarn Berry
and ClearlyDefined coordinates compare equal to their plain `name@version` equivalents.

### Bug 4 — `loadResolvedCache` stored raw identifiers as cache keys

The cache loader stored the literal backtick value as the key. A row stored as a ClearlyDefined
coordinate (`npm/npmjs/-/name/version`) or Yarn Berry format (`name@npm:version`) never matched
processor lookups that use plain `name@version`, causing a cache miss on every run for those
deps.

**Fix:** applied `coordinateToIdentifier()` normalization at both parse sites — 3-column
`prod.md`/`dev.md` rows and 2-column EXCLUDED rows.

### Bug 5 — Stale `transitive dependency` EXCLUDED entries survived when a package became direct

A package previously written to EXCLUDED as `transitive dependency` was loaded into `depsToCQ`
on subsequent runs. When that package later appeared in `dependencies`/`devDependencies`, the
`stillUnresolved` filter skipped it (already in `depsToCQ`) — it was never re-queried, never
added to `problems.md`, and never harvested.

**Fix:** added an eviction pass before the `stillUnresolved` computation. Packages cached as
`transitive dependency` that are now direct are deleted from `depsToCQ` and their EXCLUDED rows
are removed via `removeUnusedExcludes`, so they flow through the normal unresolved path.

---

## Steps to reproduce (cache miss issue)

1. Run the tool against a project with a populated `.deps/prod.md`.
2. Run again without changing `yarn.lock`.
3. Observe that the same ~1 421 API calls are made both times (same "querying N dependencies"
   count in the logs).

## Expected behavior

On the second run (no lockfile change) the log shows:

```
Cache miss: 0 dependencies to query (0 version bumps, 0 new packages).
```

and the tool finishes without making any ClearlyDefined API calls.

## Additional context

- Fixes redundant ClearlyDefined API calls.
- Prevents harvest requests being sent for transitive dependencies.
- Fixes batch POST timeout that caused consistent failures with `--harvest`.
- Fixes npm audit vulnerability (qs DoS, GHSA-q8mj-m7cp-5q26).
- Auto-removes stale EXCLUDED entries approved by ClearlyDefined.

### PR DESCRIPTION
**Pull request by olexii4**
### What does this PR do?

Several related improvements to how the tool queries and caches ClearlyDefined license data, plus automatic cleanup of stale EXCLUDED file entries.

#### 1. Dependency cache (skip already-resolved deps)

On each run the tool now reads `.deps/prod.md` and `.deps/dev.md` before calling ClearlyDefined. Any dependency whose **Resolved CQs** column is non-empty (a `clearlydefined` link or CQ number) is considered already resolved and is skipped — only new or previously unresolved dependencies are sent to the API.

Result on `eclipse-che/che-dashboard` (~1 421 total dependencies):
- **Before**: every dep queried on every run
- **After**: ~93 API queries (1 328 served from cache)

`--recheck` bypasses the cache and re-queries everything.

Key implementation details:
- `loadResolvedCache(prodMdPath, devMdPath)` reads both files, stores `{license, cq}` per identifier.
- `coordinateToIdentifier()` handles all three coordinate formats: plain `pkg@version`, Yarn Berry `pkg@npm:version`, and ClearlyDefined `npm/npmjs/…`.
- Cached entries are reconstructed in proper ClearlyDefined DEPENDENCIES format (`npm/npmjs/…, MIT, approved, clearlydefined`) so `parseDependenciesFile` can process them correctly.
- "transitive dependency" placeholder entries are excluded from the cache.
- Entries in linked-name format (`` [`pkg@version`](url) ``) are matched by an updated regex.
- Entries with an empty license column (deps sourced from EXCLUDED files, e.g. `| dep |  | ecd.che |`) are now matched correctly (`([^|]*)` instead of `([^|]+)`).

#### 2. Enriched cache-miss log

The cache-miss message now breaks down the misses by type:

```text
Cache miss: 93 dependencies to query (38 version bumps, 55 new packages).
```

- **prod / dev / prod+dev** — which dep bucket(s) (available for npm and yarn3)
- **version bumps** — package name was cached at a different version (expected after `yarn.lock` upgrades)
- **new packages** — package name not seen in the cache at all

#### 3. Configurable POST/GET timeouts

Realistic defaults; overridable via CLI or `LibraryConfig`:

| Request type | Old default | New default | CLI flag |
|---|---|---|---|
| Batch POST `/definitions` | 60 000 ms | **30 000 ms** | `--post-timeout <ms>` |
| Individual GET `/definitions/{id}` | 30 000 ms | **5 000 ms** | `--get-timeout <ms>` |

The batch POST is retried once (after 2 s) before falling back to individual GETs.

#### 4. Harvest only direct-unresolved dependencies

Previously harvest was triggered for every dep returned as `notfound` by the API — including transitive dependencies that are automatically added to `.deps/EXCLUDED/`. Transitive deps are not owned by the project and should not be harvested.

The transitive/direct split happens inside `processAndGenerateDocuments()`, which runs after the API call. Harvest now runs there too:

- `src/backends/harvest.ts` (new) — standalone `triggerHarvestAsync(identifiers, getTimeoutMs)` with two phases: await GET checks, then fire POST requests (fire-and-forget).
- `processAndGenerateDocuments()` receives an optional `harvestFn` callback called with **only the direct unresolved identifiers** (those that end up in `problems.md`). Transitive deps are excluded.
- `ClearlyDefinedBackend` no longer owns harvest logic (`enableHarvest` option and `processHarvest()` method removed).

#### 5. Persistent cache misses eliminated

Previously, certain deps were re-queried from ClearlyDefined on **every** run even though yarn.lock hadn't changed. Three root causes were found and fixed:

**Root cause A — uncategorized yarn3 deps never written to output files:**
The yarn3 lockfile parser produces `prod`, `dev`, and `all` arrays. 93 deps appeared in `all` (queried by the API and approved) but in neither `prod` nor `dev` (graph-traversal gaps for optional/transitive packages). Since `arrayToDocument` only writes deps from `prodDeps`/`devDeps`, these deps were approved but never stored in `prod.md`/`dev.md`, so never cached. Fix: uncategorized deps are appended to `devDependencies` in `yarn3-deps-info.json`.

**Root cause B — transitive deps written to EXCLUDED only with --harvest:**
The `appendTransitiveExcludes` call was gated on `--harvest`. Without that flag, transitive version-bumped deps were printed to the console but never persisted. Fix: always write transitive deps to EXCLUDED in generate mode.

**Root cause C — EXCLUDED files not read by loadResolvedCache:**
Even after Root cause B was fixed, transitive deps in EXCLUDED would still be re-queried because `loadResolvedCache` only read `prod.md`/`dev.md`. Fix: also read `EXCLUDED/prod.md` and `EXCLUDED/dev.md`; cache these entries with empty license so the chunked processor writes them as `restricted, excluded` instead of querying the API.

#### 6. EXCLUDED auto-cleanup

EXCLUDED files can accumulate stale entries over time (entries that were once unresolved or transitive but are now approved by ClearlyDefined). Two complementary mechanisms keep them clean:

**Per-run inline verification (every run):**
EXCLUDED entries whose CQ is a `[clearlydefined](…)` link and that appear in the current dependency list are sent to the ClearlyDefined batch API on every run. If the API confirms them as approved, `parseDependenciesFile` detects them as "unused excludes" and `processAndGenerateDocuments` removes them from the EXCLUDED file in the same run.

**Full recheck with `--harvest`:**
When `--harvest` is passed, `recheckAndCleanExcluded` (new: `src/helpers/recheck-excluded.ts`) re-queries every entry in `EXCLUDED/prod.md` and `EXCLUDED/dev.md` against ClearlyDefined — including orphans no longer in yarn.lock and entries with `transitive dependency` CQ. Any entry confirmed as approved is removed immediately.

### What issues does this PR fix or reference?

fixes https://github.com/eclipse-che/che/issues/23896

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -77,10 +80,19 @@ npx @eclipse-che/license-tool --jar /path/to/dash-licenses.jar
 | `--check` | Check only, do not write any files | `false` |
 | `--batch <n>` | Batch size for ClearlyDefined API requests | `500` |
 | `--harvest` | Request harvest for unresolved deps from ClearlyDefined | `false` |
+| `--recheck` | Bypass the `.deps/` cache and re-query ClearlyDefined for every dependency | `false` |
+| `--post-timeout <ms>` | Timeout for batch POST `/definitions` requests | `10000` ms |
```
Comment (akurinnoy) [inline README.md]: The default here is `10000 ms`, but `DEFAULT_POST_TIMEOUT_MS` in `clearlydefined-backend.ts` is `30000`, and the CLI `--help` text in `cli.ts` also says `30000 ms`. The `ChunkedProcessorOptions` JSDoc and `Options` JSDoc also say `10 000 ms`.

All four locations should agree on one value.

File: src/package-managers/npm/bump-deps.ts [hunk 2]
```diff
@@ -98,12 +99,17 @@ export class NpmDependencyProcessor {
       });
 
       // Process and generate documents
+      const processOptions: ProcessingOptions = { ...options };
+      if (options?.harvest) {
+        processOptions.harvestFn = (ids: string[]) => triggerHarvestAsync(ids, 5000);
+      }
```
Comment (akurinnoy) [inline src/package-managers/npm/bump-deps.ts:94]: Two things here:

1. The timeout is hard-coded to `5000` instead of reading from `options.getTimeoutMs ?? 5000`, so `--get-timeout` is ignored for harvest requests.

2. The same 4-line block appears identically in `yarn/bump-deps.ts` and `yarn3/bump-deps.ts`. Would it make sense to move this into `PackageManagerBase.checkRestrictions()` next to the existing `recheckAndCleanExcluded` integration? That way the timeout fix also lands in one place.

File: src/helpers/utils.ts [hunk 3]
```diff
@@ -456,6 +605,21 @@ export class PackageManagerUtils {
...
+      // handled by EXCLUDED files and do not need ClearlyDefined to harvest.
+      if (isHarvest && options?.harvestFn) {
+        const directUnresolved = [
+          ...stillUnresolvedProd.filter(d => isDirectPackage(d)),
+          ...stillUnresolvedDev.filter(d => isDirectPackage(d)),
+        ];
+        if (directUnresolved.length > 0) {
+          // Fire-and-forget: harvestFn is async but we don't await it so
+          // the tool output is not blocked on harvest HTTP round-trips.
+          void options.harvestFn(directUnresolved);
```
Comment (akurinnoy) [inline src/helpers/utils.ts]: `void options.harvestFn(directUnresolved)` does not attach a `.catch()`. If the promise rejects before `triggerHarvestAsync`'s internal catch handlers run, this produces an unhandled promise rejection.

```suggestion
          void options.harvestFn(directUnresolved).catch(err => {
            console.error(`Harvest failed: ${err instanceof Error ? err.message : err}`);
          });
```

File: src/helpers/recheck-excluded.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+  for (const [filePath, label] of [
+    [excludedProdPath, 'prod'],
+    [excludedDevPath, 'dev'],
+  ] as const) {
+    if (!existsSync(filePath)) continue;
+
+    const content = readFileSync(filePath, 'utf8');
+    const toVerify: Array<{ id: string; coord: string }> = [];
+    const tablePattern = /^\| `([^`]+)` \| ([^|]+) \|$/gm;
```
Comment (akurinnoy) [inline src/helpers/recheck-excluded.ts]: This regex only matches plain backtick entries (`| \`pkg@version\` | cq |`), but `loadResolvedCache` in `utils.ts` uses a broader regex that also handles linked-name entries (`` | [\`pkg@version\`](url) | ... | ``).

This means entries with linked names will be cached by `loadResolvedCache` but silently skipped during `--harvest` recheck.

Would it make sense to extract a shared regex or a `parseTableRow()` helper so both functions parse the same subset of entries?

**Review by akurinnoy — COMMENTED** [review]
Nice work on the cache system - the performance improvement (1421 → 93 API calls) is significant, and the PR description with root cause analysis is excellent.

One general concern: `coordinateToIdentifier()` and `identifierToCoordinate()` are the foundation of the entire cache - every cache lookup goes through them. They handle 3+ input formats (plain identifiers, Yarn Berry `@npm:` format, ClearlyDefined `npm/npmjs/...` coordinates) and have edge cases around scoped packages. These are pure functions that would be straightforward to unit test. Given that incorrect conversions would silently break the cache, would it be possible to add test coverage for at least these two functions before merge?

See inline comments for specific issues.

File: src/helpers/utils.ts (hunk 3 above)
Comment (akurinnoy) [reply src/helpers/utils.ts]: Addressed in aaec6d27 - `.catch()` handler added. Thanks!

File: src/helpers/recheck-excluded.ts (hunk 4 above)
Comment (akurinnoy) [reply src/helpers/recheck-excluded.ts]: Addressed in aaec6d27 - regex now handles linked-name format. Thanks!

File: src/package-managers/npm/bump-deps.ts (hunk 2 above)
Comment (akurinnoy) [reply src/package-managers/npm/bump-deps.ts:94]: Timeout fix addressed in aaec6d27 - now reads `options.getTimeoutMs ?? 5000`. Thanks!

The duplication across 3 package managers is still there though. Would it make sense to consolidate this into `PackageManagerBase` next to the existing `recheckAndCleanExcluded` integration? Not blocking, just a thought for cleaner maintenance.

File: README.md (hunk 1 above)
Comment (akurinnoy) [reply README.md]: README fixed to `30000` in aaec6d27. Thanks!

The JSDoc comments still say `10 000 ms` in three places though - see my new comment on `types.ts`.

**Review by akurinnoy — COMMENTED** [review]
Re-review after aaec6d27. All correctness bugs from the prior round have been addressed - nice work on the thorough follow-up commit.

Two remaining items (non-blocking):

1. **JSDoc POST timeout defaults** say `10 000 ms` in 3 locations (`clearlydefined-backend.ts` constructor, `chunked-processor.ts` ChunkedProcessorOptions, `types.ts` Options), but the actual default constant is `30 000 ms`. Quick fix - just update the doc comments.

2. **Test coverage for coordinate conversion** - `coordinateToIdentifier()` and `identifierToCoordinate()` are pure functions that underpin the entire cache. They handle 3 input formats with edge cases around scoped packages. Adding tests would provide confidence for future maintenance.

Design note: the harvestFn wiring block is duplicated across `npm/bump-deps.ts`, `yarn/bump-deps.ts`, and `yarn3/bump-deps.ts`. Consider consolidating into `PackageManagerBase` where `recheckAndCleanExcluded` already lives.

File: src/helpers/types.ts [hunk 5]
```diff
@@ -48,6 +48,22 @@ export interface Options {
...
+   * ClearlyDefined for every dependency, even ones already resolved.
+   * Default: false (cache is used).
+   */
+  recheck: boolean;
+  /**
+   * Timeout for ClearlyDefined batch POST /definitions requests in ms.
+   * Default: 10 000 ms.
+   */
+  postTimeoutMs?: number;
+  /**
```
Comment (akurinnoy) [inline src/helpers/types.ts:61]: The JSDoc here says `Default: 10 000 ms` but `DEFAULT_POST_TIMEOUT_MS` in `clearlydefined-backend.ts` is `30000`. The same mismatch is in `chunked-processor.ts` (ChunkedProcessorOptions JSDoc) and `clearlydefined-backend.ts` (constructor JSDoc).

Quick fix - just update the three doc comments to say `30 000 ms`.

File: src/helpers/recheck-excluded.ts (hunk 4 above)
Comment (olexii4) [reply src/helpers/recheck-excluded.ts]: I have fixed it.

File: src/helpers/utils.ts (hunk 3 above)
Comment (olexii4) [reply src/helpers/utils.ts]: I have fixed ir.

File: src/package-managers/npm/bump-deps.ts (hunk 2 above)
Comment (olexii4) [reply src/package-managers/npm/bump-deps.ts:94]: I have fixed it.

File: README.md (hunk 1 above)
Comment (olexii4) [reply README.md]: I have fixed it.

File: src/helpers/types.ts (hunk 5 above)
Comment (olexii4) [reply src/helpers/types.ts:61]: I have fixed it.

**Review by svor — APPROVED** [review]
lgtm
